File: slcm/admission/report/campus_and_program_admissions_funnel/campus_and_program_admissions_funnel.py
import frappe
from frappe import _

def execute(filters=None):
    if not filters:
        filters = {}

    columns = get_columns()
    data = get_data(filters)

    return columns, data

def get_columns():
    return [
        {"label": _("Stage"), "fieldname": "stage", "fieldtype": "Data", "width": 250},
        {"label": _("Count"), "fieldname": "count", "fieldtype": "Int", "width": 120},
        {"label": _("Conversion Rate (to next stage)"), "fieldname": "conversion", "fieldtype": "Percent", "width": 250}
    ]

def get_data(filters):
    query_filters = []
    if filters.get("admission_cycle"):
        query_filters.append(f"admission_cycle = {frappe.db.escape(filters.get('admission_cycle'))}")
    if filters.get("campus"):
        query_filters.append(f"campus = {frappe.db.escape(filters.get('campus'))}")
    if filters.get("program"):
        query_filters.append(f"program = {frappe.db.escape(filters.get('program'))}")

    where_clause = " WHERE " + " AND ".join(query_filters) if query_filters else ""
    
    # Get aggregated counts from DB - Summing everything for a single funnel
    sql = f"""
        SELECT 
            application_status, 
            COUNT(*) as count 
        FROM `tabApplicant` 
        {where_clause}
        GROUP BY application_status
    """
    results = frappe.db.sql(sql, as_dict=1)

    # Simplified status map for the whole filtered scope
    status_map = {}
    for res in results:
        status = res.application_status
        if status == "Draft":
            continue
        status_map[status] = status_map.get(status, 0) + res.count

    # Comprehensive groupings for statuses
    all_submitted = list(status_map.keys())
    all_selected = ["Selected", "Merit Selected", "Seat Selected", "Offer Issued", "Offer Accepted", "Offer Declined", "Offer Expired", "Fee Paid", "Accepted", "Enrolled"]
    all_waitlisted = ["Waitlisted", "Merit Waitlisted", "Seat Waitlisted"]
    all_rejected = ["Rejected", "Entrance Test Rejected", "Interview Rejected", "Merit Rejected", "Seat Rejected"]
    all_withdrawn = ["Withdrawn"]
    all_offered = ["Offer Issued", "Offer Accepted", "Offer Declined", "Offer Expired", "Fee Paid", "Accepted", "Enrolled"]
    all_accepted = ["Offer Accepted", "Fee Paid", "Accepted", "Enrolled"]
    all_declined = ["Offer Declined"]
    all_expired = ["Offer Expired"]
    all_fee_paid = ["Fee Paid", "Enrolled"]

    # Define all stages for the report
    report_stages = [
        {"label": _("Submitted"), "statuses": all_submitted},
        {"label": _("Selected"), "statuses": all_selected, "parent": _("Submitted")},
        {"label": _("Waitlist"), "statuses": all_waitlisted, "parent": _("Submitted")},
        {"label": _("Rejected"), "statuses": all_rejected, "parent": _("Submitted")},
        {"label": _("Withdrawn"), "statuses": all_withdrawn, "parent": _("Submitted")},
        {"label": _("Offered"), "statuses": all_offered, "parent": _("Selected")},
        {"label": _("Accepted"), "statuses": all_accepted, "parent": _("Offered")},
        {"label": _("Declined"), "statuses": all_declined, "parent": _("Offered")},
        {"label": _("Expired"), "statuses": all_expired, "parent": _("Offered")},
        {"label": _("Fee Paid"), "statuses": all_fee_paid, "parent": _("Accepted")}
    ]

    # Calculate counts for each report stage
    calculated_counts = {}
    total_submitted = 0
    for stage in report_stages:
        total = 0
        for status in stage["statuses"]:
            total += status_map.get(status, 0)
        calculated_counts[stage["label"]] = total
        if stage["label"] == _("Submitted"):
            total_submitted = total

    data = []
    for stage in report_stages:
        label = stage["label"]
        count = calculated_counts[label]
        
        conversion = 0.0
        if total_submitted > 0:
            conversion = round((count / total_submitted) * 100, 2)
        elif not stage.get("parent"): # Root stage
             conversion = 100.0
        
        data.append({
            "stage": label,
            "count": count,
            "conversion": conversion
        })

    return data

File: slcm/admission/report/campus_and_program_admissions_funnel/campus_and_program_admissions_funnel.js
frappe.query_reports["Campus and Program Admissions Funnel"] = {
    "filters": [
        {
            "fieldname": "admission_cycle",
            "label": __("Admission Cycle"),
            "fieldtype": "Link",
            "options": "Admission Cycle",
            "default": frappe.defaults.get_user_default("admission_cycle")
        },
        {
            "fieldname": "campus",
            "label": __("Campus"),
            "fieldtype": "Link",
            "options": "Campus"
        },
        {
            "fieldname": "program",
            "label": __("Programme"),
            "fieldtype": "Link",
            "options": "Programme"
        }
    ],
    "get_chart_data": function (columns, result) {
        // If only one group (Campus+Program), we can show a funnel chart
        // For multiple groups, it might be messy.
        // Let's summarize across all filtered groups for the chart
        let summary = {
            "Submitted": 0,
            "Selected": 0,
            "Waitlisted": 0,
            "Rejected": 0,
            "Offer Issued": 0,
            "Offer Accepted": 0,
            "Offer Declined": 0,
            "Offer Expired": 0,
            "Confirmation Fee Paid": 0,
        };

        result.forEach(row => {
            if (row.stage && summary.hasOwnProperty(row.stage)) {
                summary[row.stage] += row.count;
            }
        });

        if (summary.Submitted === 0) return null;

        const labels = Object.keys(summary);

        // To get multi-colored vertical bars in Frappe Charts, we use the multiple-dataset + stacking trick.
        // We use null instead of 0 to see if the chart engine omits them from the tooltip.
        const datasets = labels.map((label, i) => {
            return {
                name: label,
                values: labels.map((l, j) => (i === j ? summary[label] : null))
            };
        });

        return {
            data: {
                labels: labels,
                datasets: datasets
            },
            type: 'bar',
            height: 300,
            barOptions: {
                stacked: 1,
                space_between_bars: 20
            },
            colors: ['#007bff', '#28a745', '#ffc107', '#dc3545', '#17a2b8', '#6610f2', '#e83e8c', '#fd7e14', '#20c997']
        };
    }
};
